src/engine/renderer/tr_shader.cpp

```diff
@@ -5283,51 +5279,42 @@ static void FinishStages()
 		}
 
 		// Compute normal scale.
-		for ( int i = 0; i < 3; i++ )
+		if ( hasNormalMap )
 		{
-			if ( hasNormalMap )
+			if ( !HasNormalFormat( stage ) )
 			{
-				if ( !stage->hasNormalFormat )
-				{
-					// Please make sure the parser sets a normal format
-					// when adding a new normal map syntax.
-					ASSERT_UNREACHABLE();
-				}
-
-				if ( !stage->hasNormalScale )
-				{
-					stage->normalScale[ i ] = 1.0;
-				}
+				/* Please make sure the parser sets a normal format
+				when adding a new normal map syntax. */
+				ASSERT_UNREACHABLE();
+				SetNormalFormat( stage, dxNormalFormat );
+			}
+			
+			stage->normalScale[ 0 ] *= stage->normalFormat[ 0 ];
+			stage->normalScale[ 1 ] *= stage->normalFormat[ 1 ];
 
-				/* Because Z reconstruction is destructive on alpha channel
-				Z reconstruction is never done on normal map shipping height
-				map in alpha channel and Z is read from the file itself.
-				Those files always provide Z anyway.
+			/* Because Z reconstruction is destructive on alpha channel
+			Z reconstruction is never done on normal map shipping height
+			map in alpha channel and Z is read from the file itself.
+			Those files always provide Z anyway.
 
-				If height map is not stored in normal map alpha channel,
-				the Z component will be reconstructed from X and Y whatever
-				Z is provided by the file or not) and Z will be fine from
-				the start, so we must not apply the format translation on
-				the Z channel.
+			If height map is not stored in normal map alpha channel,
+			the Z component will be reconstructed from X and Y whatever
+			Z is provided by the file or not) and Z will be fine from
+			the start, so we must not apply the format translation on
+			the Z channel.
 
-				So this test means X and Y formats are always applied,
-				but Z format is applied only when Z is not reconstructed.
+			So this test means X and Y formats are always applied,
+			but Z format is applied only when Z is not reconstructed.
 
-				Note the XYZ format translations are not done on RGB channels
-				from the file but on the RGB channels as seen in GLSL shader,
-				there is no need to worry about DXn storing X in alpha channel.
+			The XYZ format translations are not done on RGB channels
+			from the file but on the RGB channels as seen in GLSL shader,
+			there is no need to worry about DXn storing X in alpha channel.
 
-				This way the material syntax is expected to work the same with
-				both the PNG source and the released CRN.
-				*/
-				if ( i < 2 || stage->hasHeightMapInNormalMap )
-				{
-					stage->normalScale[ i ] *= stage->normalFormat[ i ];
-				}
-			}
-			else
+			This way the material syntax is expected to work the same with
+			both the PNG source and the released CRN. */
+			if ( stage->hasHeightMapInNormalMap )
 			{
-				stage->normalScale[ i ] = 1.0;
```

Comment:
I notice the new code produces a different result for normalScale in the case of a goofy shader that has normalScale but does not have a normal map. Not sure if this matters

Reply:
It doesn't matter.

Reply:
Well, a diffuse map stage will use a flat normal map fallback if there is no normal map, so an explicit normal scale without normal map would apply to the flat normal map fallback, I'm don't think it's wrong. It's wrong for a mapper to do that, but it's not wrong for the engine to do what the mapper asked for.

One cannot change the normal format of the flat normal map fallback, but that is expected.

Reply:
The same is already true with another shader keyword allowing to modify the normals with some expression.